### DISCUSSION TITLE
Fail2Ban Rule

### DISCUSSION THREAD

**Discussion by zacc1111** [IrosTheBeggar/mStream#363] [General]
Hello,

it is possible to write a rule for fail2ban?

{"level":"warn","message":"Failed login attempt from ::ffff:80.18.19.23. Username: root","stack":{},"timestamp":"2022-04-09T01:38:36.298Z"}

Thank you.

**Comment by brazzmonkey** [comment]
I second that, a rule for fail2ban would be nice for servers that are exposed.

**Comment by rigel314** [comment]
Here's my fail2ban setup:

/etc/fail2ban/filter.d/mstream.conf:
```
[Definition]
failregex="message"\s*:\s*"Failed login attempt from <HOST>\.[^"]*
ignoreregex=
datepattern="timestamp"\s*:\s*"%%Y-%%m-%%dT%%H:%%M:%%S\.\d{3}Z"
```

/etc/fail2ban/jail.local:
```
[mstream]
enabled = true
filter = mstream
banaction = iptables
maxretry = 5
findtime = 1d
bantime = 1h
bantime.increment = true
bantime.factor = 24
bantime.maxtime = 5w
ignoreip = 192.168.0.0/24
logpath = /var/lib/mstream/logs/mstream*.log
```

root crontab:
```
*/5 * * * * /usr/bin/fail2ban-client reload mstream &> /dev/null
```


Because I'm running mStream in a docker container behind a nginx reverse proxy(running on the host), I also needed:
mStream v6.5.2+ with the config option `"trustProxy": true`
and nginx `proxy_set_header X-Forwarded-For $remote_addr;`

fail2ban uses regex and the mstream logs are json.  It's not usually a good idea to attempt to "parse" json with regex, but I think the regex I used won't be _too_ fragile.

I needed the crontab because mstream rotates logs hourly, only creating a logfile for that hour if a log event happens during that hour, and creates a newly named file every time.  Fail2ban can't automatically rescan the logpath wildcard expression, so I just used cron to tell fail2ban to reload the mstream jail every 5min - though, it would be nicer to have something using inotify to detect new files, and only then reload the jail.  But since reloading a jail doesn't clear any banned IPs, it should be fine to reload every 5min.
